[#977][WIP] Resource file list display - Folder structure phase 1

mauriel133 opened the pull request:
**Phase 1** 
deployed for testing at https://park.hydroshare.org
- [x] Create a Folder at the top (contents folder) level or within an existing folder
- [x] Delete a folder (and all the files and folders it contains)
- [x] View the contents (e.g., contained files and folders) within any folder
- [x] Delete one or more files in a folder (bulk delete)
- [x] Upload one or more files into a folder.  This should include uploads from an iRODS location
- [x] Download files individually (this function exists) 
- [x] Rename a file or folder
- [x] Move a folder into another folder (including all the files in the folder that is moved)

TODO:
- [x] Prevent page refresh switching to view mode on file delete. @hyi 
- [x] Disable file operations based on resource types. 
- [x] Test Irods functionality to handle large files with new interface.
- [x] Fix error when moving multiple files asynchronously.

Possible future improvements
- Expand file type icons.
- Delete files without having to refresh the page.
- Add 'Last Modified' field to file objects and sorting algorithms.

mauriel133 commented:
@hyi, @pkdash do you know which flag I should use to disable file operations on edit mode for resource types such as Multidimensional? Also, which operations specifically should be disabled?

pkdash commented:
@mauriel133 

> do you know which flag I should use to disable file operations on edit mode 

What prevents a user with edit permission from file operations?

mauriel133 commented:
For example, I cannot add files to a Multidimensional (NetCDF) resource I've been given edit permissions to.

pkdash commented:
Why can't you add files to any resource if you have been given edit permission on that resource?

pkdash commented:
The only time no edit is allowed if the resource has been published.

hyi commented:
@mauriel133 Look at https://github.com/hydroshare/hydroshare/blob/997-large-files-handling/hs_core/templates/pages/genericresource.html#L1914..L1994 for how adding files to existing resources are disabled for specific resource types. I believe you should be able to add the similar logic to the new file operations with folder structure.

alvacouch commented on hs_core/urls.py:
```diff
@@ -31,10 +34,12 @@
     url(r'^resource/(?P<pk>[A-z0-9]+)/files/(?P<filename>[^/]+)/$',
```
@pkdash @hyi This does not solve the file path problem. 
Consider `^resource/(P<pk>{A-z0-9]+)/files/(?P<filename>[.a-zA-Z0-9 -_/]+)/$`
What this does: a file path consists of a string of characters a-z, A-Z, 0-9, ' ', '-', '_', and '/' that ends with a '/'.
In Linux the following characters are usually not used in filenames: 

* ';', '/', '\' are shell escapes. 
* '*' is the shell glob character. 
* ',' is the glob separator 
* '$' references a shell variable. 
* '&' puts a job into the background. 
* '%' references a job. 
* '!' references a previous command. 
* '()' defines a shell array. 
* '{}' invokes special variable manipulations. 
* '`' invokes a shell command. 

For security reasons, we also need to disallow '/../' and '/./' and '//'. The regular expression engine allows me to accomplish this. 

Or, if you want, we could simply accept everything and then validate it later. E.g., `^resource/(P<pk>{A-z0-9]+)/files/(?P<filename>.+)/$` I.e., accept anything that ends with a / delimiter.

hyi replied:
@alvacouch @pkdash Our download path url allows anything with a url pattern like ```url(r'^download/(?P<path>.*)$', 'django_irods.views.download')```. Our view function does the check and serve files only when requested files are valid. That seems to be simple and works well. In that ```resource/<pk>/files/<filename>``` url for REST endpoint, if filename is only the file name without path, we should probably exclude ```/```, otherwise, I'll vote for non-restrictive url pattern for file names or file path.

mauriel133 commented:
@hyi, could you check the back end for an error when moving multiple files asynchronously?

hyi commented:
@mauriel133 I am working on this branch now reorganizing some backend view functions to get ready for REST API implementation that will come next. Will look into this error you mentioned when moving multiple files asynchronously in the meantime.

hyi commented:
@mauriel133 I have pushed my code update to the branch that reorganized backend view structures and also fixed the move files to folders error. This move error is caused by the AJAX call that pass in a target path that does not contain the source file name. Since the same backend view function is used both for move and rename, the source file name is required in the target path in order for the view function to function correctly. In other words, the target path has to include the file name as well when moving a file from a source path to target path. I have made corresponding changes in hs_file_browser.js to make it work. Let me know if you still run into issues. Also, I seem to run into an issue with file uploading. I don't seem to be able to drag files from local disk in order to add files to the resource. Not sure if that just happens in my branch or an issue you are aware of.

horsburgh commented:
@hyi and @mauriel133 - I can confirm the drag and drop issue. I tried it yesterday and it didn't work. I didn't get an error, I just didn't see the file get uploaded.

mauriel133 commented:
Thanks for testing it out. It looks like one of my last changes broke it. I'll fix it today.

mauriel133 commented:
@hyi, I'm still experiencing the error when moving multiple files. The files get moved successfully to the folder, but the POST method returns "INTERNAL SERVER ERROR".

hyi commented:
@mauriel133 Hmmm, is there any place I have missed that pass in a target path that does not contain the source file name when doing AJAX call for moving files? I only tested moving two files simultaneously by dragging two files to a sub-folder and they work as expected after I made the fix. How do you do moves which run into this error? How many files do you move all at once?

mauriel133 commented:
There was a place where it was missing: the method to drop files through drag & drop using the AI, but I already added it. I verified that the file name is being included to the path. You can test this on the latest deployment.

hyi commented:
@mauriel133 The multiple file async move server error has been fixed and the fix has been pushed to the branch. You can pull the update and test it out. Let me know if you notice any further issues.

mauriel133 commented:
@hyi, see https://github.com/hydroshare/hydroshare/pull/1491/commits/a0356d56aa50495bd80700d04847770554dcd182

There is something crashing the site when trying to merge the work right before that point. Could you have a look at what changes in that commit that might be causing the crash?

mauriel133 commented:
@hyi, disregard last message. I was able to track it down and fix it.

hyi commented:
@mauriel133 I tracked it down as well, and found you had already pushed the change when I was just about to push it. I pushed another fix for pylint issue, so expect to see a green check from Jenkins after the build is complete. Let me know if you run into further issues.

mauriel133 commented:
@hyi, thanks for taking care of that.

For Model Instance resources it has been requested that the zip files automatically unzip upon upload. Should I reuse the unzip method in the front end and automatically call it after the file has been uploaded, or is this something that is better to do as a back end check while the is being uploaded?

hyi commented:
@mauriel133 @mohamedmorsy Let's not do this automatic unzip for model instance resources for this first phase. We can add this functionality at the back end after this is deployed.

pkdash commented:
I will do the review of the backend code. @hyi Can you review of the front-end code?

hyi commented:
@pkdash @mauriel133 Yes, I will review front end code.

pkdash commented:
@hyi Are there any unit tests written for this work? Looking at all 25 files that have been changed I could not find any.

hyi commented:
@pkdash No unit tests have been written yet, but I have planned to address unit tests together with federation zone-related functionality unit tests recorded in issue https://github.com/hydroshare/hydroshare/issues/1292, since all backend functionalities implemented for folder operation involve iRODS file operations.

pkdash commented:
@hyi I see the last successful build of the branch has code coverage of 59% which is reducing the coverage by 1% point what we have now for the development branch. I would strongly recommend that we write unit tests for such large PR.

pkdash commented:
@hyi Also it will be easier for me to review the backend code if I go through the unit test code first.

hyi commented:
@pkdash @mjstealey is very close in getting irods docker containers ready into our development infrastructure. As soon as that is ready, I'll work on writing unit tests for these iRODS-related operations. Right now, we don't have that ready, and we will have to have the coverage lowered a bit during this short interim until we have iRODS docker container ready for doing iRODS backend related unit tests.

hyi commented:
@pkdash Just do you best to code review for now without unit tests. I will make sure unit tests are covered when iRODS docker containers are ready and make sure the new code all pass unit tests.

pkdash commented:
@zhiyuli and @shawncrawley It seems there is quite a bit of front-end code in this PR. If you both can review the front-end code that would be great.

mauriel133 commented:
Deployment has been moved to https://playground.hydroshare.org/ for testing purposes.

mauriel133 commented:
>  I was unable to drag and drop upload files into a folder that I created within a resource. It doesn't throw an error. It says the progress is complete, but the folder is empty. A refresh doesn't matter. Ha ha - then I navigate back to the root directory and I see that the files got uploaded to the root directory!!!! Need to fix this so that the files go into the folder where I dragged them.

This is our end goal, but at this point we only have functionality to upload files to the root folder. We will aim to have this ready for phase 2. @hyi, please confirm.
 



>  Deleting a file from the root directory made the system go crazy - I get the spinny wheel on the browser tab, a spinning icon in the folder area, and I think I got booted out of edit mode. This might just be the slow internet connection I'm on. It looks like the file did get deleted. Someone else should test this while not on a crappy slow internet connection.

Some of that might have been indeed caused by the slow connection, but it is true that the page refreshes after a file is deleted. This is necessary because deleting files might change metadata which needs to be reflected through page refresh. We plan to address this on phase 2 as well and make it consistent with the rest of the functionality.

mauriel133 commented:
> I drag and drop a text file in. I can right click on the file once uploaded, and the option to "open" is there, but not functional. Can we disable non-functional options? If the plan is to leave them disabled until the next phase, then I say just leave them.

@horsburgh, didn't the 'open' option appear disabled when you right clicked the file? If so, are you suggesting to hide it?

![image](https://cloud.githubusercontent.com/assets/2448568/19610992/a44c8de2-979c-11e6-8904-882d72810e08.png)

hyi commented:
Re: This is our end goal, but at this point we only have functionality to upload files to the root folder. We will aim to have this ready for phase 2. @hyi, please confirm.
@mauriel133 @horsburgh Yes, @mauriel133 is correct. We only allow uploading to the root folder at this point for phase 1 and will aim to have it ready for phase 2. Is it possible to disable uploading to a subfolder for phase 1 until we have that functionality ready? @mauriel133

mauriel133 commented:
As files are being uploaded there is a message informing about where they are going. I think disabling uploading inside sub-folders hinders the usability. I would advice to leave it as it is until phase 2. @horsburgh, let me know how we should proceed. 

![image](https://cloud.githubusercontent.com/assets/2448568/19611186/b8258732-979d-11e6-9ae3-9922e61a8379.png)

hyi commented:
@mauriel133 I have tested functionality of adding files from iRODS including from a federated zone on beta which was deployed by @mjstealey a couple of days ago for this branch. They all work as expected in terms of functionality. A minor issue was that it reverts to view mode after adding a file from iRODS which I have fixed and pushed the fix to github.  However, the user interface display has some alignment issues after the user signed into an iRODS server as shown below. 

![after_sign_in_from_irods](https://cloud.githubusercontent.com/assets/476302/19611955/0914793c-97b2-11e6-8199-a7f2f68a6d49.png)

The user interface display used to be like the following:

![irods_before](https://cloud.githubusercontent.com/assets/476302/19611987/2d9ec78a-97b2-11e6-82da-96276602307d.png)

Is this something you can easily fix?

mauriel133 commented:
I don't have a way of testing iRODS functionality in my local instance right now. But, by the looks of it, it might be easy to fix by setting up some margins in CSS. Could you take care of it?

hyi commented:
@mauriel133 OK. I'll take care of it then.

horsburgh commented:
@mauriel133 - the message on upload was obviously lost on me because I thought it just wasn't working.  If our only option for now is to upload to the root folder, then the user should be taken to the root folder when the upload finishes so they can see the files they just uploaded. But, I think this going to potentially confuse people. 

And, yes - I am suggesting hiding the "open" option on right click on a file until we can support it.

mauriel133 commented:
> The "select all" button is a plus sign, which might confuse users who might think that button is for adding files. Can we use a different icon? I've seen a couple that have multiple stacked boxes with a check mark in the top one.

@horsburgh, I changed the Select All button to this: 
![image](https://cloud.githubusercontent.com/assets/2448568/19613151/066d2850-97a8-11e6-823f-80a74eb27e51.png)

I couldn't find the exact one you were thinking of, but let me know if you can spot a better candidate. Here is the list of icons we have available for now: http://fontawesome.io/icons/

> @mauriel133 - the message on upload was obviously lost on me because I thought it just wasn't working. If our only option for now is to upload to the root folder, then the user should be taken to the root folder when the upload finishes so they can see the files they just uploaded. But, I think this going to potentially confuse people.

I changed the behavior so the browser loads the 'data/contents' folder after upload. Please confirm if the upload message is not showing up as you test.

These changes have been deployed to https://park.hydroshare.org

horsburgh commented:
@mauriel133 - how about this one: http://fontawesome.io/icon/check-square/

pkdash commented on hs_core/hydroshare/utils.py:
```diff
@@ -313,7 +313,8 @@ def get_resource_file_name_and_extension(res_file):
     """
     Gets the file name and extension of the specified resource file
```
Needs update to reflect that it now returns 3 things.

pkdash commented on hs_core/hydroshare/utils.py:
```diff
@@ -323,16 +324,35 @@ def get_resource_file_name_and_extension(res_file):
+    elif res_file.fed_resource_file_name_or_path:
+        f_url = res_file.fed_resource_file_name_or_path
+    else:
+        f_url = ''
```
Is this case possible ever?

hyi replied:
In normal cases, it will not be possible. But if something goes awry, it could be possible, which we actually encountered during initial backend-frontend integration. In any case, it is always good and will not hurt to include an ```else``` phrase for just-in-case precautions.

pkdash commented on hs_core/hydroshare/utils.py:
```diff
@@ -754,6 +774,20 @@ def resource_file_add_process(resource, files, user, extract_metadata=False,
 
 
+def create_empty_contents_directory(resource):
+    if resource:
```
Is this necessary?

hyi replied:
Yes, I think so. It's always nice to make sure the input object is not empty before doing anything.

pkdash replied:
Since this function is internally used from another function probably, I don't see the need for this check. Just a thought.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+import os
+
+from django.http import HttpResponse
+from rest_framework.exceptions import NotFound, PermissionDenied
```
Import ordering. Better to put this one after all jango imports

hyi replied:
@pkdash In fact, django.http is the only django import here, and the next import is django_irods module, so no change of import order is needed here.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+    """
+    res_id = request.POST['res_id']
+    if res_id is None:
+        return HttpResponse(status=400)
```
Should set the content-type to json

hyi replied:
OK. Will do.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+                                   needed_permission=ACTION_TO_AUTHORIZE.VIEW_RESOURCE)
+    except (NotFound, PermissionDenied):
+        # return permission defined response
+        return HttpResponse(status=403)
```
As above.

pkdash commented on hs_core/hydroshare/utils.py:
```diff
@@ -754,6 +774,20 @@ def resource_file_add_process(resource, files, user, extract_metadata=False,
+def create_empty_contents_directory(resource):
+    if resource:
+        res_id = resource.short_id
+        if resource.resource_federation_path:
```
I think it would be better to use a syntax like this so that we don't have to check where the resource is stored (federated or not) all the time:

```istorage = resource.get_istorage()```

hyi replied:
I'll see what I can do with this.

pkdash commented on hs_core/hydroshare/utils.py:
```diff
@@ -754,6 +774,20 @@ def resource_file_add_process(resource, files, user, extract_metadata=False,
+        res_id = resource.short_id
+        if resource.resource_federation_path:
+            istorage = IrodsStorage('federated')
+            res_contents_dir = '{}/{}/data/contents'.format(resource.resource_federation_path,
```
Same kind of logic as above for this too.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+
+    store_path = request.POST['store_path']
+    if store_path is None:
+        return HttpResponse(status=400)
```
As above.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+        return HttpResponse(status=400)
+    store_path = str(store_path).strip()
+    if not store_path:
+        return HttpResponse(status=400)
```
As above.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+    """
+    res_id = request.POST['res_id']
+    if res_id is None:
+        return HttpResponse(status=400)
```
Better to use either JsonRespone or set the content type to json.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+    except (NotFound, PermissionDenied):
+        # return permission defined response
+        return HttpResponse(status=403)
+    input_coll_path = request.POST['input_coll_path']
```
This can raise exception, right?. Should you be using ```request.POST.get('input_coll_path', None)``` instead?

hyi replied:
Agreed using ```request.POST.get('input_coll_path', None)``` is better. Will make changes.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+    unzipped, and remove_original_zip has a value of "true" or "false" (default is "true")
+    indicating whether original zip file will be deleted after unzipping.
+    """
+    res_id = request.POST['res_id']
```
Same as above.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+    except (NotFound, PermissionDenied):
+        # return permission defined response
+        return HttpResponse(status=403)
+    zip_with_rel_path = request.POST['zip_with_rel_path']
```
Could causes exception  as pointed out above.

pkdash commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+    where folder_path is the relative path for the new folder to be created under
+    res_id collection/directory.
+    """
+    res_id = request.POST['res_id']
```
Could cause exception.

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,251 @@ def show_relations_section(res_obj):
+
+def link_irods_file_to_django(resource, filename, size=0):
+    # link the newly created zip file to Django resource model
+    if resource:
```
Is this check necessary?

hyi replied:
Yes, this function is called recursively for folder hierarchy, and it is always a good thing to check to make sure operations are only done for input object is not empty.

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,251 @@ def show_relations_section(res_obj):
+                                             resource_file=filename).exists():
+                ResourceFile.objects.create(content_object=resource,
+                                            resource_file=filename)
+
```
Should we be also create metadata 'format' element when adding this file?

hyi replied:
Probably, will double check and add it if needed.

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,251 @@ def show_relations_section(res_obj):
+     default is empty meaning nothing is excluded.
+    :return:
+    """
+    if resource and istorage and foldername:
```
Do we need this check?

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,251 @@ def show_relations_section(res_obj):
+    :return:
+    """
+    if resource and istorage and foldername:
+        fed_path = resource.resource_federation_path
```
Not needed.

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,251 @@ def show_relations_section(res_obj):
+    """
+    resource = hydroshare.utils.get_resource_by_shortkey(res_id)
+    if resource.resource_federation_path:
+        istorage = IrodsStorage('federated')
```
Again if we can use ```istorage = resource.get_istorage()```

pkdash commented:
@hyi I am now done with 1st round review of backend code for this PR.

hyi commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,251 @@ def show_relations_section(res_obj):
+                # a new ResourceFile with new file associated that replace the old one
+                res_file_obj[0].delete()
+                ResourceFile.objects.create(content_object=resource, resource_file=tgt_name)
+
```
@pkdash Yes, adding new files are not changed - still using the old existing view functions. The only change is the GUI part with no backend being changed.

hyi commented on hs_core/views/resource_folder_hierarchy.py:
```diff
@@ -0,0 +1,313 @@
+                    f_pk = f.pk
+                    f_url = get_resource_file_url(f)
+                    break
+            files.append({'name': fname, 'size': size, 'type': mtype, 'pk': f_pk, 'url': f_url})
```
The front-end does not check file extension but uses the mime type returned from back-end directly, which is why I added mimetype as part of file attribute to return to front-end here.

mauriel133 commented:
Team, 

Test your resources with this new interface at https://park.hydroshare.org/ and raise any concerns or suggestions you might have in this thread.

@hyi, I'm done implementing front end changes at this point and the code is ready for review.

zhiyuli commented:
@mauriel133 
Got a error when deleting *.shp and *.shx files at the same time from a geo-feature feature resource
![image](https://cloud.githubusercontent.com/assets/10070599/19667086/f8a5b762-9a0b-11e6-89e3-00f0a417e56c.png)

zhiyuli commented:
After deleting one required file (shp or shx or dbf) from geo-feature res, all existing res files get deleted. Then, failed to add a new unzipped shapefile (contains shp, shx, dbf and other accessory files) and no error or warning showed up.

hyi commented:
@zhiyuli I have changed delete_multiple_files() to ignore this specific ObjectDoesNotExist exception when deleting multiple files one after the other to accommodate this specific case. I have also updated park.hydroshare.org with updated code. @zhiyuli Can you test again to see if it is sufficient to address this specific case?

zhiyuli commented:
@hyi  a quick test on park indicated the deletion action in feature res is working as expected now.

hyi commented:
@pkdash I have addressed all your comments and added unit tests and fixed a couple of issues that were revealed by unit test cases. So the backend is ready for your next round of review now.

pkdash commented:
@hyi That's the beauty of unit tests - as they help eliminate bugs. I will try to do another round of review by tomorrow. Thanks.

zhiyuli commented:
@pkdash @hyi @mauriel133 
based on my test, the issue of adding shapefile still exists. It still uploads each component of shapefile separately instead of uploading them together.

hyi commented:
@zhiyuli I think if you do not do "drag and drop" to add files, but rather do file browsing to add files as you used to, then it should add files all together as it used to, right? I think the way @mauriel133 implemented drag and drop for adding files is to add each file via a separate ajax call, so they are added one by one rather than all together if you do "drag and drop," is that true, @mauriel133? If so, I'll defer this to you and @mauriel133 to work out a good way to handle this special case.

zhiyuli commented:
@hyi @mauriel133 
I just tried it again. Looks like this link works differently than before. It uploads each file separately as doing drag-and-drop. 
![image](https://cloud.githubusercontent.com/assets/10070599/19701918/b4d9e332-9aba-11e6-8e4b-5c3a2cb3a1fd.png)

mauriel133 commented:
@hyi, I think you are right. It has to do with how dropzone.js uploads files. I will try to fix through the configurations.

hyi commented on theme/templates/resource-landing-page/modals.html:
```diff
@@ -1711,9 +1711,97 @@ <h4 class="modal-title" id="delete-funding-agency">Delete Funding Agency</h4>
+                        <label>Name *</label>
+                        <div class="input-group">
+                            <input id="txtName" class="form-control modal-only-required">
+                            {% if cm|resource_type == "Time Series" or cm|resource_type == "Multidimensional (NetCDF)" %}
```
@mauriel133 It looks like there is no "rename" action available for netCDF resource type, so this conditional check (at least for netCDF resource type) is not needed, right?

hyi commented on theme/static/js/resourceLandingAjax.js:
```diff
@@ -470,4 +470,176 @@ function get_user_info_ajax_submit(url, obj) {
+                    $(".fb-upload-caption").toggleClass("hidden", true);
+                }
+                else {
+
```
@mauriel133 better to remove this extra line if you have other things to change so as to commit this together with other changes needed.

hyi commented:
@mauriel133 @pkdash I have gone through the front end code with a couple of very minor comments. It looks good to me. A +1 from me after the issue with geofeature resource type multiple file add/upload all at once is resolved by configuring dropzone as @mauriel133 mentioned above.

mauriel133 commented:
@hyi, @pkdash I'm working on fixing the multiple upload problem and I have submitted partial work. Avoid re-deploying on the VM until I'm done, hopefully soon.

pkdash commented on hs_core/tests/api/native/test_resource_file_folder_operations.py:
```diff
@@ -0,0 +1,175 @@
+        self.assertIn(self.new_res.short_id + '/data/contents/sub_test_dir/file2.txt',
+                      updated_res_file_names,
+                      msg='resource does not contain unzipped file file2.txt')
+
```
Should you be also checking that the file file3_new.txt exists after unzip?

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,278 @@ def show_relations_section(res_obj):
+    :param tgt_name: the file or folder full path name to be renamed to
+    :return:
+    """
+    if resource:
```
Why check only one parameter and not the rest? In fact, I don't think this check is necessary.

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,278 @@ def show_relations_section(res_obj):
+    """
+    resource = hydroshare.utils.get_resource_by_shortkey(res_id)
+    if resource.resource_federation_path:
+        istorage = IrodsStorage('federated')
```
Should this be ```istorage = resource.get_istorage()```?

pkdash commented on hs_core/views/utils.py:
```diff
@@ -404,3 +404,278 @@ def show_relations_section(res_obj):
+    """
+    resource = hydroshare.utils.get_resource_by_shortkey(res_id)
+    if resource.resource_federation_path:
+        istorage = IrodsStorage('federated')
```
Same as above.

pkdash commented:
@hyi Have some minor comments in my 2nd round of review. Glad to see the unit tests.

mauriel133 commented:
@hyi, I've made some changes and made sure  that the files are being sent in one request. However, there are some validations failing in the back end. Could you debug and see what's wrong?

Also, I have an event in my dropzone object listening for errors. Would be good if the ajax call can return an error that the event can listen to. 

```
 // An error occured. Receives the errorMessage as second parameter and if the error was due to the XMLHttpRequest the xhr object as third.
                this.on("error", function (error, errorMessage, xhr) {
                    $("#fb-alerts .upload-failed-alert").remove();
                    $("#fbContainmentWrapper").toggleClass("glow-blue", false);

                    $("#fb-alerts").append(
                            '<div class="alert alert-danger alert-dismissible upload-failed-alert" role="alert">' +
                                '<button type="button" class="close" data-dismiss="alert" aria-label="Close">' +
                                    '<span aria-hidden="true">&times;</span></button>' +
                                '<div>' +
                                    '<strong>File Upload Failed</strong>'+
                                '</div>'+
                                '<div>'+
                                    '<span>' + errorMessage + '</span>' +
                                '</div>'+
                            '</div>').fadeIn(200);
                });
```

mauriel133 commented:
@hyi, I fixed the multiple upload issue now. I tested with Shape File as well.

mauriel133 commented on theme/static/js/hs_file_browser.js:
```diff
@@ -0,0 +1,969 @@
+    while (path){
+        var folder = path.substr(path.lastIndexOf("/") + 1, path.length);
+        var currentPath = path;
+        path = path.substr(0, path.lastIndexOf("/"));
```
@hyi, the 'path' variable is being used to extract the value for the 'folder' variable in each iteration.
Although it is not expected to be empty, I added a base case conditional in case an error causes it to be empty.

hyi commented:
@mauriel133 I just tested multiple upload implementation and found although the multiple files are added to the resource without issues, the geo-feature resource specific metadata elements are not extracted. I will look into it and see if I can fix it.

hyi commented:
@mauriel133 The existing backend view function add_file_to_resource() assumes the request is not ajax, but a page request. It looks like from here: https://github.com/hydroshare/hydroshare/blob/997-large-files-handling/hs_core/templates/pages/genericresource.html#L390-L413 that the request for adding files to resources are indeed a form submit, which used to refresh the whole page after files are added. However, with your dropzone folder implementation, it seems it acts as an AJAX request without refreshing the page. I am not sure where AJAX is done, probably somewhere implicitly in dropzone, but the backend view function still assumes the request is a page request and return a response ```HttpResponseRedirect(request.META['HTTP_REFERER'])```. I guess dropzone encapsulates this form in its container, and when the backend view function returns that ```HttpResponseRedirect()```, it refreshes the container without refreshing the whole page. Also this backend view functions set request.session['validation_error'] for these validation errors to be displayed on the top of the page via django template context variables when the page is refreshed. Now that it does not refresh the whole page, but only refresh the container, I am not sure whether these template context variables can be leveraged or not. In short, I am a bit confused on how this backend view function (which assumes the request is not AJAX but a page request) is used and whether we need to change this backend view function to handle AJAX request due to your dropzone folder implementation.

mauriel133 commented:
@hyi, Does the problem concern updating metadata or showing the error messages after the AJAX call?

hyi commented:
@mauriel133 The problem I found is with adding multiple files for geographic feature resource type. If I upload the same set of multiple files when creating the resource, the resource specific metadata is extracted as expected. However, if I create an empty feature resource type, then add the same set of files, the resource specific metadata is not extracted. This issue led me to look into that add_resource_file() backend view function, then I realized it is an existing function that assume the request is not AJAX, but a page request, although it does not restrict to only non-AJAX requests. You also asked to handle errors after the AJAX call. Since this view function just set errors on request.session which is then passed as context variable to template page to display after page is refreshed, it appears to me it needs to be changed to return error to your AJAX call, but then I get confused on how you actually call this function via AJAX. This function is still called via a form submit action. Can you confirm whether this ```/hsapi/_internal/{{ cm.short_id }}/add-file-to-resource/``` action is still submitted via AJAX? If so, to be able to display errors in your AJAX call back, what should I return to the front end? A JSON response with {```'error': error_message} or something like that?

hyi commented:
@mauriel133 @zhiyuli After debugging, I found the metadata are indeed extracted after adding multiple files to geo-feature resource. After refreshing the page, I see those extracted resource-specific metadata. So I believe if you can refresh the page, things will work/look as expected.

mauriel133 commented:
@hyi, I added success and error listeners  that I can use to display the response. Can we change the response of the method to a JSON object rather than a page refresh?

mauriel133 commented:
Does that mean we should change the code to call a page refresh after adding a file?

hyi commented:
@mauriel133 To supplement my previous comment, at least for specific resource types, after adding files to existing resource, the whole page should be refreshed since new metadata is added to the resource which only shows up after refreshing the page. It'd be good to auto-refresh the page so users don't have to do it in order to see updated content on the page. This also justify why the original add_file_to_resource() view function expect the request to be a regular page request rather than an AJAX request. I think if you can refresh the whole landing page after adding files at least for specific resource type, it should be good.

hyi commented:
@mauriel133 Yes, I think an automatic page refresh after adding files sounds good, which will ensure the whole page is updated after files are added.

hyi commented:
@pkdash I have addressed all your code review comments. See if you can give a +1 to backend code now. Thanks.

hyi commented:
@mauriel133 On second thought, can you make add files to resource call a regular page request as it used to rather than an AJAX call? If so, the errors will be displayed on the page upon page refresh without extra coding.

mauriel133 commented:
Dropzone works by doing an AJAX call to the file upload. I don't think I can change it to do a POST request instead. I can call a page refresh right after upload finishes, but that would not carry the context variables.

hyi commented:
@mauriel133 OK. Then I will change that function to return a JSON response with success and error messages then. In the meantime, can you add page refresh call? Then we can do integration test after we both are done.

mauriel133 commented:
Can you give me a list of resource types for which page refresh is needed after file upload?

hyi commented:
@mauriel133 All specific resource types need a page refresh due to metadata extraction from the files being added. I believe only generic resource does not need a page refresh. @pkdash Can you confirm?

mauriel133 commented:
Done, I added refresh for non-generic resource types after files are uploaded. I also added methods to restore the previous scroll position in order to make the refresh less intrusive.

hyi commented:
@mauriel133 Great. I'll take care of the backend changes and test to make sure it works.

hyi commented:
@mauriel133 I have pushed backend changes needed for adding files to resource. If you can update the VM with latest code for people to do another round of quick testing as needed, it'd be good. @pkdash see if you can +1 for the backend code. For front-end code, a +1 from me. @zhiyuli @shawncrawley It'd be great if you could code review front-end code as well when you have time, but if not, I think the front-end is good to go.

zhiyuli commented:
@mauriel133 @hyi 
I tested feature res type on park and it worked as expected regarding removing and adding multiple files. Good job!
The front-end code looks good enough to me.  +1.

pkdash commented:
@hyi +1 for the backend code.

hyi commented:
+1 for both backend and frontend code, so merging.
